### DISCUSSION TITLE
Library environment broken (?) in formula test phase, in CI only

### DISCUSSION THREAD

**Discussion by lelegard** [Homebrew/discussions#4738] [Writing Formulae/Casks]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

Hi,

What are the specific characteristics of the running environment in the formula test phase during CI tests, following a pull request in homebrew-core?

Here is the situation I have. I fixed an existing formula for `libtomcrypt`. This is not a new version. This is a fix of an existing formula that I didn't author. In short, `libtomcrypt` was unable to perform asymmetric cryptography in Homebrew because of a missing dependency on `libtommath` (and a few missing symbols in the `make` command).

I fixed it on my macOS13 arm64 system. I fixed the formula, reinstalled `--from-source` using the locally fixed formula. It works:
- `brew test libtomcrypt` passes. It basically runs the test program which was build during the installation phase.
- Applications work and asymmetric cryptography algorithms are now available.

So far, so good, I submitted a [pull request](https://github.com/Homebrew/homebrew-core/pull/140370) for the fixed formula.

Here comes the problem: the test phase of the macOS11-arm64 fails in the CI (and tests on other platforms were cancelled). The same `brew test libtomcrypt` (which passes on my system) fails in the CI with this error:
~~~
==> Testing libtomcrypt
==> ./test
dyld: Library not loaded: @loader_path/../../../../../opt/libtommath/lib/libtommath.1.dylib
    Referenced from: /private/tmp/libtomcrypt-test-20230824-29211-1af4hzi/./test
    Reason: image not found
~~~

The library `libtommath.1.dylib` is not found. This library is provided by `libtommath` (note: "math", not "crypt"), the new dependency I added in the formula.

Note that the build of `libtomcrypt` succeeded. Since the build uses `libtommath`, the library was there at the time of the build.

However, the very same library is no longer found in the test phase.

I am a bit surprised by the referenced path `@loader_path/../../../../../opt/libtommath/lib`. It this expected? This path is obviously the reason for the error.

The same test command works on my system.

Additional information:

- For some reason, the test program is built during the build phase and installed as `share/libtomcrypt/test`. This executable tests all supported cryptographic algorithms and can be run at any time after installation. Why? No idea.
- The test program only depends on the `libtommath` library, the dependency I added. Before the fix, this dependency was not present, which explains why it worked.
~~~
$ otool -L test
test:
	/opt/homebrew/opt/libtommath/lib/libtommath.1.dylib (compatibility version 4.0.0, current version 4.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
~~~
- The referenced path is hardcoded during the build. On my system:
~~~
$ otool -l test
...
Load command 14
          cmd LC_LOAD_DYLIB
      cmdsize 80
         name /opt/homebrew/opt/libtommath/lib/libtommath.1.dylib (offset 24)
   time stamp 2 Thu Jan  1 01:00:02 1970
      current version 4.0.0
compatibility version 4.0.0
...
~~~

What should I update in the formula to make sure that the test application correctly runs in the CI?

**Comment by ZhongRuoyu** [comment]
Thanks for the pull request. Please see review at https://github.com/Homebrew/homebrew-core/pull/140370#pullrequestreview-1595414286.
